hail/python/hail/ir/renderer.py

```diff
@@ -134,3 +135,563 @@ def __call__(self, x: 'Renderable'):
+
+                if lift_type == 'value':
+                    visited = lift_to_frame.visited
+                    name = lift_to_frame.lifted_lets[id(child)]
+                elif lift_type == 'agg':
+                    visited = lift_to_frame.agg_visited
+                    name = lift_to_frame.agg_lifted_lets[id(child)]
+                else:
+                    visited = lift_to_frame.scan_visited
+                    name = lift_to_frame.scan_lifted_lets[id(child)]
+
+                frame.builder.append(f'(Ref {name})')
+
+                if id(child) in visited:
+                    continue
+
+                if lift_type == 'value':
+                    child_builder = [f'(Let {name} ']
+                elif lift_type == 'agg':
+                    child_builder = [f'(AggLet {name} False ']
+                else:
+                    child_builder = [f'(AggLet {name} True ']
+
+                if id(child) in memo:
+                    child_builder.append(memo[id(child)])
+                    child_builder.append(' ')
+                    lift_to_frame.let_bodies.append(child_builder)
+                    continue
+
+                visited[id(child)] = child
+
+                new_state = frame.make_child_frame(self.renderer, binding_sites,
+                                                   child_builder, bindings_stack,
+                                                   child_min_binding_depth,
+                                                   child_min_value_binding_depth,
+                                                   child_scan_scope,
+                                                   child_depth)
+                new_state.lift_to_frame = lift_to_frame
+                stack.append(new_state)
+                continue
+
+            if id(child) in memo:
+                frame.builder.append(memo[id(child)])
+                continue
+
+            new_state = frame.make_child_frame(self.renderer, binding_sites,
+                                               frame.builder, bindings_stack,
+                                               child_min_binding_depth,
+                                               child_min_value_binding_depth,
+                                               child_scan_scope,
+                                               child_depth)
+            stack.append(new_state)
+            continue
+
+    BindingsStackFrame = namedtuple(
+        'BindingsStackFrame',
+        'depth lifted_lets agg_lifted_lets scan_lifted_lets visited agg_visited'
+        ' scan_visited let_bodies')
+
+    class StackFrame:
+        __slots__ = ['node', 'children', 'min_binding_depth',
+                     'min_value_binding_depth', 'scan_scope', 'depth',
+                     'lift_to_frame', 'insert_lets', 'builder', 'child_idx']
+
+        def __init__(self,
+                     node: Renderable,
+                     children: Sequence[Renderable],
+                     builder: Sequence[str],
```

Comment:
Should be a List, not Sequence. You use `append` all over the place, and that's a List function, not part of the Sequence interfae.

Reply:
Ah right. Though it looks like the right abc is MutableSequence.

Reply:
Ah right. Though it looks like the right abc is MutableSequence.